pkg/build/builder/source.go

```diff
@@ -104,35 +109,43 @@ func fetchSource(dockerClient DockerClient, dir string, build *api.Build, urlTim
 // remote repository failed to authenticate.
 // Since this is calling the 'git' binary, the proxy settings should be
 // available for this command.
-func checkRemoteGit(gitClient GitClient, url string, timeout time.Duration) error {
-	glog.V(4).Infof("git ls-remote --heads %s", url)
+func checkRemoteGit(gitClient GitClient, url string, initialTimeout time.Duration) error {
 
 	var (
 		out    string
 		errOut string
 		err    error
 	)
 
-	out, errOut, err = gitClient.TimedListRemote(timeout, url, "--heads")
-	if _, ok := err.(*git.TimeoutError); err != nil && ok {
-		return fmt.Errorf("timeout while waiting for remote repository %q", url)
-	}
-
-	if len(out) != 0 {
-		glog.V(4).Infof(out)
-	}
-	if len(errOut) != 0 {
-		glog.V(4).Infof(errOut)
+	timeout := initialTimeout
+	for {
+		glog.V(4).Infof("git ls-remote --heads %s", url)
+		out, errOut, err = gitClient.TimedListRemote(timeout, url, "--heads")
+		if len(out) != 0 {
+			glog.V(4).Infof(out)
+		}
+		if len(errOut) != 0 {
+			glog.V(4).Infof(errOut)
+		}
+		if err != nil {
+			if _, ok := err.(*git.TimeoutError); ok {
+				nextTimeout := timeout * timeoutIncrementFactor
+				glog.Infof("WARNING: timed out waiting %s for git server, will wait %s", timeout, nextTimeout)
```

Comment:
Don't display the first timeout duration, it's redundant.  Just indicate the next interval.